src/pocketmine/network/upnp/UPnP.php

```diff
@@ -19,63 +19,201 @@
+
 declare(strict_types=1);
 
 /**
- * UPnP port forwarding support. Only for Windows
+ * UPnP port forwarding support.
  */
 namespace pocketmine\network\upnp;
 
 use pocketmine\utils\Internet;
-use pocketmine\utils\Utils;
-use function class_exists;
-use function is_object;
+use SimpleXMLElement;
+use function socket_create;
+use function socket_set_option;
+use function socket_recvfrom;
+use function socket_sendto;
+use function trim;
+use function strlen;
+use function preg_match;
+use function parse_url;
+use const AF_INET;
+use const SOCK_DGRAM;
+use const SOL_UDP;
+use const IPPROTO_IP;
+use const IP_MULTICAST_IF;
+use const IP_MULTICAST_LOOP;
+use const IP_MULTICAST_TTL;
+use const SOL_SOCKET;
+use const SO_RCVTIMEO;
 
 abstract class UPnP{
+	/** @var string */
+	private static $serviceURL;
+
+	/**
+	 * Search the device node with given type.
+	 *
+	 * @param \SimpleXMLElement $device
+	 * @param string $deviceType
+	 * @param \SimpleXMLElement $foundDevice reference parameter
+	 *
+	 * @return bool
+	 */
+	private static function getChildDevice(\SimpleXMLElement $device, string $deviceType, &$foundDevice) : bool{
+		foreach($device->deviceList->device as $child){
+			if((string)$child->deviceType === $deviceType){
+				$foundDevice = $child;
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Search the service node with given type.
+	 *
+	 * @param \SimpleXMLElement $device
+	 * @param string $serviceType
+	 * @param \SimpleXMLElement $foundService reference parameter
+	 *
+	 * @return bool
+	 */
+	private static function getChildService(\SimpleXMLElement $device, string $serviceType, &$foundService) : bool{
+		foreach($device->serviceList->service as $child){
+			if((string)$child->serviceType === $serviceType){
+				$foundService = $child;
+				return true;
+			}
+		}
+		return false;
+	}
+	
+	public static function setup() : void{
+		$socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+		socket_set_option($socket, SOL_SOCKET, SO_RCVTIMEO, array("sec"=>3,"usec"=>0));
+		$contents =
+			"M-SEARCH * HTTP/1.1\r\n".
+			"MX: 2\r\n".
+			"HOST: 239.255.255.250:1900\r\n".
+			"MAN: \"ssdp:discover\"\r\n".
+			"ST: upnp:rootdevice\r\n\r\n";
+		socket_sendto($socket, $contents, strlen($contents), 0, "239.255.255.250", 1900);
+		socket_recvfrom($socket, $buffer, 1024, 0, $responseHost, $responsePort);
+
+		if(!$buffer || !preg_match('/location\s*:\s*(.+)\n/i', $buffer, $matches)){
+			throw new \RuntimeException("Unable to find the router. Ensure that network discovery is enabled in Control Panel.");
+		}
+		$location = trim($matches[1]);
+		$url = parse_url($location);
+		$response = Internet::getURL($location, 3, [], $err, $headers, $httpCode);
+		if(!$response){
+			throw new \RuntimeException("Unable to access XML: {$err}");
+		}
+		if($httpCode !== 200){
+			throw new \RuntimeException("Unable to access XML: {$response}");
+		}
+		$root = new \SimpleXMLElement($response);
+		$device = $root->device;
```

Comment:
This might not exist, or it might be of the wrong type.

Reply:
So should I check the existence of `device` by using `isset()`?

Reply:
Yes, and the type.

Reply:
I found `SimpleXMLElement::xpath`, which enable us to explore inside a XML without executing null check or type check in each XML node (All we need will be just checking `empty($root->xpath(...))` is true or not, I think.)
I'm now rewriting the code by using this xpath so it will take a little longer to commit.

Reply:
`empty()` has some issues with phpstan. It's advised to use `is_null()` where applicable for more accurate analysis.

Reply:
`is_null()` is a useless function, and `isset()` has the same problems as `empty()`, so I'm a little confused what the point is here.

Reply:
To be clear, the reason `empty()` is banned by PHPStan is because it disguises an `isset() and != null` check behind a misleading name.

Reply:
I mean, since xpath() returns false if some error occurs and returns array(0) if nothing is found, I tried to use empty() to check these two cases.
Now I'm checking them by `$xpathResult === false` and `count($xpathResult) === 0`

Reply:
Yes, don't use weak type checks. That goes for `!$result` on functions that return `something|false` as well.